apple/internal/aspects/swift_dynamic_framework_aspect.bzl

```diff
@@ -0,0 +1,144 @@
+""",
+)
+
+def _swift_target_for_dep(dep):
+    """Returns the target for which the dependency was compiled.
+
+    This is really hacky, but there's no easy way to acquire the Apple CPU for which the target was
+    built. One option would be to let this aspect propagate transitively through deps and have
+    another provider that propagates the CPU, but the model there gets a bit more complicated to
+    follow. With this approach, we avoid propagating the aspect transitively as well.
+
+    This should be cleaned up when b/141931700 is fixed (adding support for ctx.rule.split_attr).
+    """
+    for action in dep.actions:
+        if action.mnemonic == "SwiftCompile":
+            target_found = False
+            for arg in action.argv:
+                if target_found:
+                    return arg
+                if arg == "-target":
+                    target_found = True
+    fail("error: Expected at least one Swift compilation action for target {}.".format(dep.label))
+
+def _swift_arch_for_dep(dep):
+    """Returns the architecture for which the dependency was built."""
+    target = _swift_target_for_dep(dep)
+    return target.split("-", 1)[0]
+
+def _modulemap_contents(module_name):
+    """Returns the contents for the modulemap file for the framework."""
+    return """\
+framework module {module_name} {{
+  header "{module_name}.h"
+  requires objc
+}}
+""".format(module_name = module_name)
+
+def _swift_dynamic_framework_aspect_impl(target, ctx):
+    """Aspect implementation for Swift dynamic framework support."""
+    if not hasattr(ctx.rule.attr, "deps"):
+        return []
+    
+    swiftdeps = [x for x in ctx.rule.attr.deps if SwiftInfo in x]
+    ccinfos = [x for x in ctx.rule.attr.deps if CcInfo in x]
+    # If there are no Swift dependencies, return nothing.
+    if not swiftdeps:
+        return []
+
+    # Collect all relevant artifacts for Swift dynamic framework generation.
+    module_name = None
+    generated_header = None
+    swiftdocs = {}
+    swiftmodules = {}
+    modulemap_file = None
+    for dep in swiftdeps:
+        swiftinfo = dep[SwiftInfo]
+
+        if not module_name:
+            module_name = swiftinfo.module_name
+
+        arch = _swift_arch_for_dep(dep)
+
+        # Collect the interface artifacts.
+        if swiftinfo.transitive_generated_headers:
+            if not generated_header:
+                generated_header = swiftinfo.transitive_generated_headers.to_list().pop()
+
+        swiftdocs[arch] = swiftinfo.transitive_swiftdocs.to_list().pop()
+        swiftmodules[arch] = swiftinfo.transitive_swiftmodules.to_list().pop()
+        modulemap_file = ctx.actions.declare_file("{}_file.modulemap".format(module_name))
+        ctx.actions.write(modulemap_file, _modulemap_contents(module_name))
+
+    for dep in ccinfos:
+        headers = dep[CcInfo].compilation_context.headers.to_list()
+        if headers:
+            generated_header = headers.pop(0)
```

Comment:
Since it's been a while since I've dealt with mixed source frameworks, what is the use case here? Could `generated_header` have been set above by swift, and if so, is it valid to override it here?

Reply:
The generated header is needed for Xcode to use the .framework file, which is output by this rule. The CcInfo provider was the only way I could see for how to get that file. In a prior implementation, the generated header field was included with the SwiftInfo provider, but I think it got deprecated. Setting the generated_header variable above is no longer relevant and I will remove it in my next push.